wrapper/src/main/java/software/amazon/jdbc/plugin/IamAuthMultithreadedConnectionPlugin.java

```diff
@@ -0,0 +1,194 @@
+  static final ConcurrentHashMap<String, Future<?>> cacheThead = new ConcurrentHashMap<>();
+  static final ExecutorService executorService = Executors.newCachedThreadPool(
+      r -> {
+        final Thread prefetchThread = new Thread(r);
+        prefetchThread.setDaemon(true);
+        return prefetchThread;
+      });
+
+  public IamAuthMultithreadedConnectionPlugin(
+      @NonNull PluginService pluginService) {
+    super(pluginService);
+  }
+
+  @Override
+  protected Connection connectInternal(String driverProtocol, HostSpec hostSpec, Properties props,
+      JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+    final long startTime = System.nanoTime();
+
+    if (StringUtils.isNullOrEmpty(PropertyDefinition.USER.getString(props))) {
+      throw new SQLException(PropertyDefinition.USER.name + " is null or empty.");
+    }
+
+    String host = hostSpec.getHost();
+    if (!StringUtils.isNullOrEmpty(IAM_HOST.getString(props))) {
+      host = IAM_HOST.getString(props);
+    }
+
+    int port = hostSpec.getPort();
+    if (!hostSpec.isPortSpecified()) {
+      if (StringUtils.isNullOrEmpty(IAM_DEFAULT_PORT.getString(props))) {
+        port = this.pluginService.getDialect().getDefaultPort();
+      } else {
+        port = IAM_DEFAULT_PORT.getInteger(props);
+        if (port <= 0) {
+          throw new IllegalArgumentException(
+              Messages.get(
+                  "IamAuthConnectionPlugin.invalidPort",
+                  new Object[] {port}));
+        }
+      }
+    }
+
+    final String iamRegion = IAM_REGION.getString(props);
+    final Region region = StringUtils.isNullOrEmpty(iamRegion)
+        ? getRdsRegion(host)
+        : Region.of(iamRegion);
+
+    final int tokenExpirationSec = IAM_EXPIRATION.getInteger(props);
+
+    final String cacheKey = getCacheKey(
+        PropertyDefinition.USER.getString(props),
+        host,
+        port,
+        region);
+    final TokenInfo tokenInfo = tokenCache.get(cacheKey);
+    final boolean isCachedToken = tokenInfo != null;
+
+    if (isCachedToken) {
+      LOGGER.finest(
+          () -> Messages.get(
+              "IamAuthConnectionPlugin.useCachedIamToken",
+              new Object[] {tokenInfo.getToken()}));
+      PropertyDefinition.PASSWORD.set(props, tokenInfo.getToken());
+    } else {
+      final String token = generateAuthenticationToken(
+          hostSpec,
+          props,
+          host,
+          port,
+          region);
+      LOGGER.finest(
+          () -> Messages.get(
+              "IamAuthConnectionPlugin.generatedNewIamToken",
+              new Object[] {token}));
+      PropertyDefinition.PASSWORD.set(props, token);
+      tokenCache.put(
+          cacheKey,
+          new TokenInfo(token, Instant.now().plus(tokenExpirationSec, ChronoUnit.SECONDS)));
+      final String threadHost = host;
+      final int threadPort = port;
+      cacheThead.put(
+          cacheKey,
+          executorService.submit(() -> {
```

Comment:
Is this adding a new thread for every connection? If so, that wouldn't be ideal. We should rather have a single thread for the entire instance that is spawned on instantiation that loops every prefetch time (OR some time less than iamExpiration) and goes through every key in cacheThread on each loop and checks for expirations).

Then when the destructor for this class is called the "while (true)" should be replaced with a "while (isDestroyed)" that is set to true by the destructor of this class to clean up the thread so it doesn't leak.

Lastly, when doing while loops within a thread, you should sleep otherwise this will hammer the CPU time of the client. I would expect to sleep < tokenExpirationSec and that the trigger to call generateAuthenticationToken is also less than when the token actually expires. If you made those the same there would be a possible race condition so I think this needs to wake up just a bit more frequent than expirations could occur.

What the above does:
1. Uses less resources because only a single thread.
2. Ensures we aren't leaking threads on close.
3. Ensure we aren't wasting CPU cycles (because we would sleep the thread at some time just < expiration).

Reply:
1. To clarify, the IAM plugin uses 1 thread per cache key, not for every connection. This cache key is the connection URL. So if the user decides to use an instance endpoint, then monitoring thread creates 1 thread for that instance endpoint. If users makes more connections to the same endpoint, the plugin uses the same thread.

2, 3. Before we make any conclusion, we're investigating through a bug identified that is impacting performance measurements right now.